python/pipeline/odor.py

```diff
@@ -0,0 +1,268 @@
+    odorant_description = ''    : varchar(2048)
+    """
+    contents = [
+        ['Empty', '0', '0', '0', '']
+    ]
+
+
+@schema
+class OdorSolution(dj.Manual):
+    definition = """ # Solutions made for olfactory experiments.
+    -> Odorant
+    concentration               : decimal(2,2)         # Dilution fraction
+    solution_date               : date                 # date solution was created (YYYY-MM-DD)
+    ---
+    solution_notes = ''         : varchar(2048)
+    """
+
+
+@schema
+class OdorSession(dj.Manual):
+    definition = """ # Single session of olfactory experiment.
+    -> mice.Mice
+    odor_session                : smallint unsigned    # session index for the mouse
+    ---
+    odor_path                   : varchar(2048)        # folder where h5 files for olfaction are stored
+    odor_session_date           : date                 # date session was recorded (YYYY-MM-DD)
+    time_loaded = null          : time
+    """
+
+
+@schema
+class OdorConfig(dj.Manual):
+    definition = """ # Configuration of solutions for each olfactory experiment. One entry for each channel.
+    -> OdorSession
+    channel                     : tinyint unsigned     # channel number of assigned solution
+    ---
+    -> OdorSolution
+    """
+
+
+@schema
+class OdorRecording(dj.Manual):
+    definition = """ # Single scan/recording within an olfactory session.
+    -> OdorSession
+    recording_idx               : smallint unsigned    # recording index for session
+    ---
+    filename                    : varchar(255)
+    """
+
+
+@schema
+class OdorTrials(dj.Imported):
+    definition = """ # Single presentation of olfactory stimulus. One entry for each ON channel.
+    -> OdorRecording
+    trial_idx                   : smallint unsigned    # trial index for recording
+    channel                     : tinyint unsigned     # channel number used
+    ---
+    trial_start_time            : float                # start of trial on olfactory clock (seconds)
+    trial_end_time              : float                # end of trial on olfactory clock (seconds)
+    """
+
+    def convert_valves(num):
+        """ Converts decimal number encoding valve states into boolean array
+        :param num: int. Decimal number encoding valve state in binary
+        :returns: numpy array of booleans. Array encodes valve state (True=OPEN) starting with the smallest valve number
+            ex. decimal(10) -> [False, True, False, True] -> 2nd and 4th valve open.
+        """
+
+        # Format integer into binary, then convert each 0/1 into boolean
+        # Finally, reverse list to make first index encode first valve
+        valve_states = [bool(int(d)) for d in format(int(num), 'b')][::-1]
+        return np.array(valve_states)
+
+    def make(self, key):
+
+        print(f'Populating trials for {key}')
+
+        # Get olfactory h5 path and filename
+        olfactory_path = (OdorSession & key).fetch1('odor_path')
```

Comment:
https://github.com/cajal/pipeline/blob/master/python/pipeline/experiment.py#L467 In experiment.session we have the attribute local_filenames_as_wildcard that takes care of finding the path so we don't have to rewrite it every time, maybe you can have something similar. Definitely not needed , if you think this works better for this pipeline.